tests/filecheck/transforms/lift-arith-to-linalg.mlir

```diff
@@ -1,15 +1,66 @@
-// RUN: xdsl-opt %s -p lift-arith-to-linalg | filecheck %s
+// RUN: xdsl-opt %s -p "lift-arith-to-linalg{strategy=from_partially_bufferized}" | filecheck %s
 
 builtin.module {
+// CHECK: builtin.module {
+
+  func.func @a() {
     %t0, %t1, %t2, %t3 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, tensor<8xf32>)
     %0 = arith.addf %t0, %t1 : tensor<8xf32>
     %1 = arith.subf %0, %t2 : tensor<8xf32>
     %2 = arith.mulf %1, %t3 : tensor<8xf32>
-}
+    func.return
+  }
 
-// CHECK-NEXT: builtin.module {
+// CHECK-NEXT: func.func @a() {
 // CHECK-NEXT:   %t0, %t1, %t2, %t3 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, tensor<8xf32>)
 // CHECK-NEXT:   %0 = linalg.add ins(%t0, %t1 : tensor<8xf32>, tensor<8xf32>) outs(%t0 : tensor<8xf32>) -> tensor<8xf32>
 // CHECK-NEXT:   %1 = linalg.sub ins(%0, %t2 : tensor<8xf32>, tensor<8xf32>) outs(%0 : tensor<8xf32>) -> tensor<8xf32>
 // CHECK-NEXT:   %2 = linalg.mul ins(%1, %t3 : tensor<8xf32>, tensor<8xf32>) outs(%1 : tensor<8xf32>) -> tensor<8xf32>
+// CHECK-NEXT:   func.return
+// CHECK-NEXT: }
+
+
+  func.func @choose_correct_tensor() {
+    // `to_tensor`s that are not writable should not be used in `outs`
+    %t0, %t1, %t2, %m0 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, memref<8xf32>)
+    %0 = bufferization.to_tensor %m0 restrict : memref<8xf32>
+    %1 = bufferization.to_tensor %m0 restrict writable : memref<8xf32>
+    %2 = bufferization.to_tensor %m0 restrict : memref<8xf32>
+    %3 = arith.addf %t0, %t1 : tensor<8xf32>
+    %4 = arith.mulf %3, %t2 : tensor<8xf32>
+    func.return
+  }
+
+// CHECK-NEXT: func.func @choose_correct_tensor() {
+// CHECK-NEXT:   %t0, %t1, %t2, %m0 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, memref<8xf32>)
+// CHECK-NEXT:   %0 = bufferization.to_tensor %m0 restrict : memref<8xf32>
+// CHECK-NEXT:   %1 = bufferization.to_tensor %m0 restrict writable : memref<8xf32>
+// CHECK-NEXT:   %2 = bufferization.to_tensor %m0 restrict : memref<8xf32>
+// CHECK-NEXT:   %3 = linalg.add ins(%t0, %t1 : tensor<8xf32>, tensor<8xf32>) outs(%1 : tensor<8xf32>) -> tensor<8xf32>
+// CHECK-NEXT:   %4 = linalg.mul ins(%3, %t2 : tensor<8xf32>, tensor<8xf32>) outs(%3 : tensor<8xf32>) -> tensor<8xf32>
+// CHECK-NEXT:   func.return
+// CHECK-NEXT: }
```

Comment:
This example worries me, where do you check that `%2` is not used after `%1` has been written to?

Reply:
I'm sorry to be annoying, but I can't shake the feeling that this is just the wrong place to add the bufferization logic. Conceptually, arith operations create a new tensor, they're not in dps, so it makes sense for them to alloc a new tensor. A pass that helps bufferization and tries to look for a memref to reuse should be separated and have these kinds of corner cases checked separately from the arith and linalg dialects

Reply:
It looks for a tensor to re-use rather than a memref.

Reply:
> This example worries me, where do you check that `%2` is not used after `%1` has been written to?

This is actually why I included it in the filecheck.

Reply:
I see that in this example it's not used, but I'm missing the bit of the code that examines the number of uses of the tensor in the pass. I agree that eliminating empty tensors is important, all I'm saying is that this custom logic would be better placed in a different pass that would eliminate them whether they were produced by lifting arith or any other method.

Reply:
It sounds like what you are proposing might be a canonicalisation pass, but this would be a canonicalisation pass that would have to be done on linalg and would differ from upstream and may also have wider impact than we want. I do not think of allocating empty tensors here as a viable option tbh.

Reply:
I don't see why this would be a canonicalization pass, it could just be its own pass with a single rewrite that implements exactly the logic added here, but on any empty tensor in the code. That way it would not be coupled to either arith or linalg, and could be tested separately.